bus-mapping/src/evm/opcodes/returndatacopy.rs

```diff
@@ -27,18 +27,21 @@ impl Opcode for Returndatacopy {
         let memory = &mut call_ctx.memory;
         let length = size.as_usize();
         if length != 0 {
+            let minimal_length = dest_offset.as_usize() + length;
+            memory.extend_at_least(minimal_length);
+
             let mem_starts = dest_offset.as_usize();
             let mem_ends = mem_starts + length;
             let data_starts = offset.as_usize();
             let data_ends = data_starts + length;
-            let minimal_length = dest_offset.as_usize() + length;
+
             if data_ends <= return_data.len() {
-                memory.extend_at_least(minimal_length);
                 memory[mem_starts..mem_ends].copy_from_slice(&return_data[data_starts..data_ends]);
-            } else {
-                assert_eq!(geth_steps.len(), 1);
-                // if overflows this opcode would fails current context, so
-                // there is no more steps.
+            } else if let Some(actual_length) = return_data.len().checked_sub(data_starts) {
```

Comment:
Will this branch get chance to be executed? Copying return data out of bound should lead to exception and should be catch by `get_step_err` then handled by its own exception execution state IIUC.

Reply:
`returndatacopy` should behave like `calldatacopy` thus, out of bounds access is valid and just plain zeroes.

Reply:
But from the implementation, it seems `RETURNDATACOPY` copies out of bound is specially handled:

```go
// copied from https://github.com/ethereum/go-ethereum/blob/master/core/vm/instructions.go#L322-L342
func opReturnDataCopy(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) ([]byte, error) {
	var (
		memOffset  = scope.Stack.pop()
		dataOffset = scope.Stack.pop()
		length     = scope.Stack.pop()
	)

	offset64, overflow := dataOffset.Uint64WithOverflow()
	if overflow {
		return nil, ErrReturnDataOutOfBounds
	}
	// we can reuse dataOffset now (aliasing it for clarity)
	var end = dataOffset
	end.Add(&dataOffset, &length)
	end64, overflow := end.Uint64WithOverflow()
	if overflow || uint64(len(interpreter.returnData)) < end64 {
		return nil, ErrReturnDataOutOfBounds
	}
	scope.Memory.Set(memOffset.Uint64(), length.Uint64(), interpreter.returnData[offset64:end64])
	return nil, nil
}
```

Reply:
Holy crap. There is no mention of this in the yellow paper but looks like [EIP-211](https://eips.ethereum.org/EIPS/eip-211) explicitly requires this.

Reply:
yeah that's one of easy-to-forget features of EVM 🥲